[MRG] Move creation of result object to `Optimizer` class

betatim opened the pull request:
Fixes #315

Move the call to `create_result` to `Optimizer` so that we can use the plot tools we have with ask-and-tell.

betatim commented on skopt/space/space.py:
```diff
@@ -214,6 +214,9 @@ def inverse_transform(self, Xt):
         return (self.low, self.high)
 
+    def __contains__(self, point):
+        return self.low <= point <= self.high
```
are these the right comparison operators? (We should document this in the doc string)

betatim commented on skopt/space/space.py:
```diff
@@ -284,6 +287,9 @@ def inverse_transform(self, Xt):
         return (self.low, self.high)
 
+    def __contains__(self, point):
+        return self.low <= point <= self.high
```
are these the right comparison operators? (We should document this in the doc string)

glouppe replied:
We have documented the upper bound of a `Real` dimension to be exclusive.

betatim replied:
Sorry, didn't spot it in the doc string :(

betatim commented:
As well as moving the creation of the results object I also added some bounds checking in this PR. Ready for reviewing.

glouppe commented:
Looks good to me, except the comment for the upper bound.

betatim commented:
Fixed the operators in the bounds checking and modified the tests a bit. They were using identical comparison operators for both `Integer` and `Real`. However one is inclusive and the other exclusive at the upper boundary.

betatim commented:
Mistake in my implementation somewhere or is our docstring wrong? https://travis-ci.org/scikit-optimize/scikit-optimize/jobs/203851498

betatim commented:
I think it is because [`fmin_l_bfgs_b()`](https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.optimize.fmin_l_bfgs_b.html) interprets the bounds differently (both ends inclusive) to us.

Ahh yiis: https://github.com/scipy/scipy/blob/master/scipy/optimize/lbfgsb/README

betatim commented on skopt/space/space.py:
```diff
@@ -214,15 +214,18 @@ def inverse_transform(self, Xt):
     def transformed_bounds(self):
         if self.transform_ == "normalize":
-            return 0.0, 1.0
+            return 0.0, 1.0 - np.finfo(float).eps
```
Is this the best we can do? Or should we modify the upper edge just before the call to BFGS?

betatim commented on skopt/space/space.py:
```diff
@@ -214,15 +214,19 @@ def inverse_transform(self, Xt):
+        eps = np.finfo(float).eps
         if self.transform_ == "normalize":
-            return 0.0, 1.0
+            return 0.0, 1.0 - eps
```
Is this the best we can do? Or should we modify the upper edge just before the call to BFGS?

betatim commented:
Decided to change the behaviour of `Real` to be inclusive of upper edge to be consistent with BFGS behaviour. Not sure if this adaptation of `scipy.stats.distributions.uniform` does what I think it does.

betatim commented:
`np.all(p in self.space for p in x)` can you spot the bug in this?

Time to merge this.

glouppe commented:
Looks good to me! Didnt know about `nextafter`, learned something :)
